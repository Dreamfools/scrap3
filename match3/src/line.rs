use crate::{BoardMatch, MatchColor};
use nohash_hasher::{IntMap, IsEnabled};
use smallvec::SmallVec;
use std::collections::BTreeSet;
use std::hash::{Hash, Hasher};

#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
struct MatchIndex(usize);

impl Hash for MatchIndex {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.hash(state)
    }
}

impl IsEnabled for MatchIndex {}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
struct BoardIndex(usize);
#[derive(Debug, Clone)]
pub struct LineMatcherSettings {
    /// Minimum line size
    pub line_size: usize,
    /// Minimum size of a match group
    pub min_group_size: Option<usize>,
    /// Whenever neighbouring valid groups of matching types should be merged
    /// regardless of the [line_size]
    pub merge_neighbours: bool,
}

impl LineMatcherSettings {
    pub fn new(line_size: usize, min_group_size: Option<usize>, merge_neighbours: bool) -> Self {
        Self {
            line_size,
            min_group_size,
            merge_neighbours,
        }
    }

    /// Common match-3 settings, with 3-in-a-row matches and no neighboring
    /// groups merging
    pub fn common_match3() -> Self {
        Self::new(3, None, false)
    }

    pub fn find_matches<
        'a,
        Color: MatchColor,
        Gem: AsRef<Color>,
        Line: AsRef<[usize]>,
        Neighbours: AsRef<[usize]>,
    >(
        &self,
        cells: &'a [Gem],
        lines: &'a [Line],
        neighbours: &'a [Neighbours],
    ) -> Vec<BoardMatch<Color>> {
        let state = LineMatcherState::new(self.clone(), cells, lines, neighbours);
        state.find_matches()
    }
}

#[derive(Debug)]
struct LineMatcherState<
    'a,
    Color: MatchColor,
    Gem: AsRef<Color>,
    Line: AsRef<[usize]>,
    Neighbours: AsRef<[usize]>,
> {
    settings: LineMatcherSettings,

    cells: &'a [Gem],
    lines: &'a [Line],
    neighbours: &'a [Neighbours],

    matches: Vec<Option<BoardMatch<Color>>>,
    match_board: IntMap<usize, SmallVec<[MatchIndex; 1]>>,

    match_cells_cache: Option<Vec<usize>>,
}

impl<
        'a,
        Color: MatchColor,
        Gem: AsRef<Color>,
        Line: AsRef<[usize]>,
        Neighbours: AsRef<[usize]>,
    > LineMatcherState<'a, Color, Gem, Line, Neighbours>
{
    fn new(
        settings: LineMatcherSettings,
        cells: &'a [Gem],
        lines: &'a [Line],
        neighbours: &'a [Neighbours],
    ) -> Self {
        Self {
            settings,
            cells,
            lines,
            neighbours,
            matches: Default::default(),
            match_board: Default::default(),
            match_cells_cache: None,
        }
    }

    fn find_matches(mut self) -> Vec<BoardMatch<Color>> {
        for line in self.lines {
            self.match_line(line.as_ref());
        }
        self.matches.into_iter().flatten().collect()
    }

    fn close_match(&mut self, mut group: BoardMatch<Color>) {
        if group.cells.len() < self.settings.line_size {
            group.cells.clear();
            self.match_cells_cache = Some(group.cells);
            return;
        }

        /// Check for intersection with other groups and merge them
        fn check_merge_groups_at_cell<Color: MatchColor>(
            matches: &mut [Option<BoardMatch<Color>>],
            match_board: &IntMap<usize, SmallVec<[MatchIndex; 1]>>,
            group: &mut BoardMatch<Color>,
            cell: usize,
            merge_group: &mut Option<MatchIndex>,
            groups_to_merge: &mut BTreeSet<MatchIndex>,
        ) {
            if let Some(other_groups) = match_board.get(&cell) {
                for &intersecting in other_groups {
                    // Resolve the other group ID
                    let other_group = &mut matches[intersecting.0]
                        .as_mut()
                        .expect("All dead groups should be inaccessible from the board");
                    if !other_group.color.matches(&group.color) {
                        continue;
                    }
                    if let Some(group) = merge_group {
                        if group == &intersecting {
                            continue;
                        }
                        // We already found a group to merge into, so add this matching group to merge in at a later stage
                        groups_to_merge.insert(intersecting);
                    } else {
                        // We intersect with the first matching group, so merge the current group into that one
                        other_group.cells.extend(&group.cells);
                        *merge_group = Some(intersecting);
                    }
                }
            }
        }

        let mut merge_group: Option<MatchIndex> = None;
        let mut groups_to_merge = BTreeSet::<MatchIndex>::new();

        // Check for intersection with other groups
        for i in 0..group.cells.len() {
            let cell = group.cells[i];
            check_merge_groups_at_cell(
                &mut self.matches,
                &self.match_board,
                &mut group,
                cell,
                &mut merge_group,
                &mut groups_to_merge,
            );
        }

        if merge_group.is_none() && group.cells.len() < self.settings.min_group_size.unwrap_or(0) {
            group.cells.clear();
            self.match_cells_cache = Some(group.cells);
            return;
        }

        // Check for merging with neighbours
        if self.settings.merge_neighbours {
            for i in 0..group.cells.len() {
                let cell = group.cells[i];
                for &neighbour in self.neighbours[cell].as_ref() {
                    check_merge_groups_at_cell(
                        &mut self.matches,
                        &self.match_board,
                        &mut group,
                        neighbour,
                        &mut merge_group,
                        &mut groups_to_merge,
                    );
                }
            }
        }

        if let Some(merged) = merge_group {
            for &x in &group.cells {
                let groups = self.match_board.entry(x).or_default();
                if !groups.contains(&merged) {
                    groups.push(merged)
                }
            }

            // Merge group exists, which mean the current group is already
            // merged, now just clean up the other groups
            for &other_group_idx in &groups_to_merge {
                let mut other = std::mem::take(&mut self.matches[other_group_idx.0])
                    .expect("Merge group was checked for already");
                let main = self.matches[merged.0]
                    .as_mut()
                    .expect("Merge group was checked for already");

                // Remap the cells of the other group to the main group
                for cell in &other.cells {
                    let groups = self
                        .match_board
                        .get_mut(cell)
                        .expect("Cell should already be initialized by the previous group");

                    #[cfg(debug_assertions)]
                    if groups.is_empty() {
                        panic!(
                            "Something gone extremely wrong\ngroup: {:?}\nmatch_groups: {:?}\ngroup: {:?}",
                            merged, self.match_board, group.cells
                        );
                    }

                    let mut remove_after = groups.len() - 1;
                    for (i, x) in groups.iter_mut().enumerate() {
                        if x == &other_group_idx {
                            *x = merged;
                            remove_after = i;
                        }
                    }
                    for i in ((remove_after + 1)..groups.len()).rev() {
                        if groups[i] == merged {
                            groups.remove(i);
                        }
                    }
                }
                main.cells.append(&mut other.cells);
            }

            #[cfg(debug_assertions)]
            for (i, cell) in self.match_board.iter() {
                for dead_group in &groups_to_merge {
                    for existing in cell {
                        debug_assert!(
                            existing != dead_group,
                            "Should not have consumed group `{}` on the board (at position `{}`)",
                            dead_group.0,
                            i,
                        );
                    }
                }
            }

            #[cfg(debug_assertions)]
            {
                for (idx, groups) in self.match_board.iter() {
                    for i in 0..(groups.len() - 1) {
                        if groups[(i + 1)..].contains(&groups[i]) {
                            panic!("Cell {} has bad groups composition: {:?}", idx, groups)
                        }
                    }
                }
            }

            group.cells.clear();
            self.match_cells_cache = Some(group.cells);
        } else {
            debug_assert!(
                groups_to_merge.is_empty(),
                "should not have extra merge groups when the main merge group is None"
            );

            let index = MatchIndex(self.matches.len());

            for &cell in &group.cells {
                self.match_board.entry(cell).or_default().push(index);
            }

            self.matches.push(Some(group));
        }
    }

    fn match_line(&mut self, line: &[usize]) {
        let mut current_match: Option<BoardMatch<Color>> = None;
        let mut was_wildcard = false;
        for i in 0..line.len() {
            let pos = line[i];
            let gem = self.cells[pos].as_ref();
            let can_start_match = gem.can_start_match();
            let can_be_matched = !gem.hint_is_unmatchable();

            if let Some(ref mut match_group) = current_match {
                if !match_group.color.matches(gem) {
                    let group = std::mem::take(&mut current_match)
                        .expect("Should have a match group to close");
                    self.close_match(group);
                } else {
                    match_group.cells.push(pos);
                }
            }

            if current_match.is_none() && can_start_match && can_be_matched {
                let mut group = BoardMatch::<Color>::new(
                    gem.clone(),
                    std::mem::take(&mut self.match_cells_cache).unwrap_or_default(),
                );

                if was_wildcard {
                    for i in (0..i).rev() {
                        let back_pos = line[i];
                        let back_gem = self.cells[back_pos].as_ref();
                        if !group.color.matches(back_gem) {
                            break;
                        }
                        group.cells.insert(0, back_pos);
                    }
                }

                group.cells.push(pos);

                current_match = Some(group);
            }
            was_wildcard = !can_start_match && can_be_matched;
        }

        if let Some(group) = current_match {
            self.close_match(group);
        }
    }
}
