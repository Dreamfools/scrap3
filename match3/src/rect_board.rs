use crate::line::LineMatcherSettings;
use crate::{BoardMatch, Gem, MatchColor};
use ndshape::{RuntimeShape, Shape};
use std::fmt::{Debug, Formatter};
use std::marker::PhantomData;
use std::sync::Arc;

#[derive(Clone)]
pub struct RectBoard<T: Gem> {
    pub shape: RuntimeShape<usize, 2>,
    pub board: Vec<T>,
    pub lines: Arc<Vec<Vec<usize>>>,
    pub neighbours: Arc<Vec<Vec<usize>>>,
}

impl<T: Debug + Gem> Debug for RectBoard<T> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("CharBoard")
            .field("width", &self.shape.as_array()[0])
            .field("height", &self.shape.as_array()[1])
            .field("board", &self.board)
            .finish()
    }
}

impl<T: Copy + Gem> RectBoard<T> {
    pub fn from_element(width: usize, height: usize, filler: impl Into<T>) -> Self {
        let board = vec![filler.into(); width * height];
        Self::new(width, height, board)
    }
}

impl<T: Gem> RectBoard<T> {
    pub fn from_fn(width: usize, height: usize, filler: impl Fn(usize) -> T) -> Self {
        let board = (0..(width * height)).map(filler).collect();
        Self::new(width, height, board)
    }

    pub fn new(width: usize, height: usize, board: Vec<T>) -> Self {
        let shape = RuntimeShape::<usize, 2>::new([width, height]);
        let mut lines: Vec<Vec<usize>> = vec![];
        let mut neighbours: Vec<Vec<usize>> = vec![];
        for y in 0..height {
            lines.push((0..width).map(|x| shape.linearize([x, y])).collect());
        }
        for x in 0..width {
            lines.push((0..height).map(|y| shape.linearize([x, y])).collect());
        }

        for i in 0..shape.size() {
            let [x, y] = shape.delinearize(i);
            let mut cell_neighbours = vec![];
            if x > 0 {
                cell_neighbours.push(shape.linearize([x - 1, y]));
            }
            if x < width - 1 {
                cell_neighbours.push(shape.linearize([x + 1, y]));
            }
            if y > 0 {
                cell_neighbours.push(shape.linearize([x, y - 1]));
            }
            if y < height - 1 {
                cell_neighbours.push(shape.linearize([x, y + 1]));
            }
            neighbours.push(cell_neighbours);
        }

        Self {
            shape,
            board,
            lines: Arc::new(lines),
            neighbours: Arc::new(neighbours),
        }
    }

    pub fn find_matches_linear(&self, settings: &LineMatcherSettings) -> Vec<BoardMatch<T::Color>> {
        settings.find_matches(&self.board, &self.lines, &self.neighbours)
    }

    pub fn width(&self) -> usize {
        self.shape.as_array()[0]
    }

    pub fn height(&self) -> usize {
        self.shape.as_array()[1]
    }
}
