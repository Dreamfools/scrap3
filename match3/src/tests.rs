use crate::line::{LineMatcherSettings, LinesBoardMatch};
use crate::rect_board::RectBoard;
use crate::MatchColor;
use colored::Colorize;
use insta::assert_snapshot;
use itertools::Itertools;
use ndshape::RuntimeShape;
use nohash_hasher::IntSet;
use rstest::rstest;
use serde::Deserialize;
use std::fmt::Debug;
use std::path::{Path, PathBuf};

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
struct CharGem(char);

impl MatchColor for CharGem {
    fn matches(&self, other: &Self) -> bool {
        self.0 == other.0 || self.0 == '*' || other.0 == '*'
    }

    fn can_start_match(&self) -> bool {
        self.0 != '*' && self.0 != '-'
    }

    fn hint_is_unmatchable(&self) -> bool {
        self.0 == '-'
    }
}

impl AsRef<CharGem> for CharGem {
    fn as_ref(&self) -> &CharGem {
        self
    }
}

impl From<char> for CharGem {
    fn from(c: char) -> Self {
        CharGem(c)
    }
}

type CharBoard = RectBoard<CharGem, CharGem>;
type S = LineMatcherSettings;

fn board_from_str(board: &str) -> CharBoard {
    let lines: Vec<&str> = board.lines().map(|l| l.trim()).collect();
    let height = lines.len();
    let width = lines[0].len();
    let shape = RuntimeShape::<usize, 2>::new([width, height]);
    let board = lines
        .into_iter()
        .enumerate()
        .flat_map(|(i, line)| {
            assert_eq!(
                line.len(),
                width,
                "All lines must have the same length, at line {i}"
            );
            line.chars().map(CharGem::from)
        })
        .collect();
    CharBoard::new(shape, board)
}

#[derive(Deserialize)]
struct MatchSettings {
    line_size: Option<usize>,
    min_group_size: Option<usize>,
    merge_neighbours: Option<bool>,
}

fn check_path(prefix: &str, path: impl AsRef<Path>) {
    let path = path.as_ref();
    let test = std::fs::read_to_string(path).unwrap();
    let mut lines = test.lines().peekable();
    let name = lines.next().expect("Should have name").trim();
    let mut settings_variations: Vec<(&'static str, S)> = vec![];
    if let Some(peek) = lines.peek() {
        if peek.starts_with('{') {
            let mut match_settings = S::common_match3();
            let settings = serde_json5::from_str::<MatchSettings>(peek.trim()).unwrap();

            match_settings.line_size = settings.line_size.unwrap_or(match_settings.line_size);
            match_settings.min_group_size =
                settings.min_group_size.or(match_settings.min_group_size);
            match_settings.merge_neighbours = settings
                .merge_neighbours
                .unwrap_or(match_settings.merge_neighbours);

            if settings.merge_neighbours.is_none() {
                match_settings.merge_neighbours = false;
                let mut cloned = match_settings.clone();
                settings_variations.push(("(with merge_neighbours = false)", match_settings));
                cloned.merge_neighbours = true;
                settings_variations.push(("(with merge_neighbours = true)", cloned));
            } else {
                settings_variations.push(("", match_settings));
            }

            let _ = lines.next();
        } else {
            settings_variations.push(("(with merge_neighbours = false)", {
                let mut s = S::common_match3();
                s.merge_neighbours = false;
                s
            }));
            settings_variations.push(("(with merge_neighbours = true)", {
                let mut s = S::common_match3();
                s.merge_neighbours = true;
                s
            }));
        }
    } else {
        panic!("Board is empty")
    }
    let board = board_from_str(&lines.join("\n"));

    let variants = settings_variations
        .into_iter()
        .map(|(settings_name, settings)| {
            let board = board.clone();
            let matches = board.find_matches(settings);
            (
                settings_name,
                visualize_snapshot(name.to_string(), board, matches, false),
            )
        })
        .collect::<Vec<(&'static str, String)>>();

    for i in 1..variants.len() {
        assert_eq!(
            variants[i].1, variants[0].1,
            "Got different results with {} and {}",
            variants[i].0, variants[0].0
        );
    }

    let file_name = path
        .file_name()
        .expect("Cases should have a file name")
        .to_string_lossy()
        .to_string();

    assert_snapshot!(format!("{}__{}", prefix, file_name), variants[0].1);
}

fn color_char(c: char) -> String {
    match c {
        'r' => "r".red().on_red().to_string(),
        'g' => "g".bright_green().on_bright_green().to_string(),
        'b' => "b".bright_blue().on_bright_blue().to_string(),
        'w' => "w".bright_white().on_bright_white().to_string(),
        'y' => "y".bright_yellow().on_bright_yellow().to_string(),
        'p' => "p".bright_purple().on_bright_purple().to_string(),
        'o' => "o".yellow().on_yellow().to_string(),
        '-' => "-".black().on_black().to_string(),
        _ => c.to_string(),
    }
}

fn display(board: &CharBoard, colored: bool) -> String {
    let body = board
        .board
        .iter()
        .map(|c| {
            if colored {
                color_char(c.0)
            } else {
                c.0.to_string()
            }
        })
        .chunks(board.width())
        .into_iter()
        .map(|mut line| format!("│{}│", line.join("")))
        .join("\n");

    format!(
        "╭{:─>width$}╮\n{}\n╰{:─>width$}╯",
        "",
        body,
        "",
        width = board.width()
    )
}

fn visualize_snapshot(
    name: String,
    mut board: CharBoard,
    mut matches: Vec<LinesBoardMatch<CharGem>>,
    colored: bool,
) -> String {
    let mut text = format!("{name}\nBoard: \n");

    text += &display(&board, colored);

    // Result order is not guaranteed, so we sort it
    matches.sort_by(|a, b| a.cells.cmp(&b.cells).then(a.color.0.cmp(&b.color.0)));

    if matches.is_empty() {
        text += "\nNo matches"
    } else {
        let mut to_remove: IntSet<usize> = Default::default();
        for (i, m) in matches.iter().enumerate() {
            let mut cloned = board.clone();
            for (i, x) in cloned.board.iter_mut().enumerate() {
                if !m.cells.contains(&i) {
                    *x = CharGem(' ');
                }
            }
            text += &format!(
                "\nMatch #{i} - {}:\n{}",
                m.color.0,
                display(&cloned, colored)
            );
            to_remove.extend(&m.cells);
        }
        for (i, x) in board.board.iter_mut().enumerate() {
            if to_remove.contains(&i) {
                *x = CharGem(' ');
            }
        }
        text += &format!("\nRemaining cells:\n{}", display(&board, colored));
    }
    text
}

#[rstest]
pub fn common_line3_file_tests(#[files("src/cases/common/*.txt")] path: PathBuf) {
    check_path("common", path);
}
// #[rstest]
// pub fn wildcard_line3_file_tests(#[files("src/cases/wildcard/*.txt")] path: PathBuf) {
//     check_path("common", path);
// }

#[test]
pub fn test_dev() {
    let b = board_from_str("rrr\nr-r\nrrr");
    let matches = b.find_matches(S::common_match3());
    visualize_snapshot("DEV".to_string(), b, matches, false);
}
