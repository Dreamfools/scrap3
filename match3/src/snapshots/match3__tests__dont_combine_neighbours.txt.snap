---
source: match3/src/tests.rs
expression: "visualize_snapshot(name, board, matches, false)"
---
Should not combine neighboring groups when merge_neighbours=false
Board: 
╭─────╮
│rrr--│
│--rrr│
╰─────╯
Match #0:
╭─────╮
│rrr  │
│     │
╰─────╯
Match #1:
╭─────╮
│     │
│  rrr│
╰─────╯
Remaining cells:
╭─────╮
│   --│
│--   │
╰─────╯
